File: algorithms/2_medium/L/02958_length_of_longest_subarray_with_at_most_k_frequency/hashmap.cpp
class Solution {
public:
    static int maxSubarrayLength(const vector<int>& nums, const int k) {
        assert(k >= 1 && k <= std::ssize(nums) && nums.size() <= 100'000);
        const size_t len = nums.size();
        auto freqs = unordered_map<int, int>();
        size_t lo = 0;
        size_t max = 0;
        for (size_t hi = 0; hi < len; hi += 1) {
            freqs[nums[hi]] += 1;
            while (freqs[nums[hi]] > k) {
                freqs[nums[lo]] -= 1;
                lo += 1;
            }
            max = std::max(max, hi - lo + 1);
        }

        return static_cast<int>(max);
    }
};

File: algorithms/2_medium/L/02958_length_of_longest_subarray_with_at_most_k_frequency/hashmap.rs
use std::collections::HashMap;

impl Solution {
    pub fn max_subarray_length(nums: Vec<i32>, k: i32) -> i32 {
        assert!(k >= 1 && (k as usize) <= nums.len() && nums.len() <= 100_000);
        let k = k as u32;
        let len = nums.len();
        let mut freqs: HashMap<i32, u32> = HashMap::new();
        let mut lo = 0_usize;
        let mut max = 0_usize;
        for hi in 0..len {
            *freqs.entry(nums[hi]).or_insert(0) += 1;
            while freqs[&nums[hi]] > k {
                *freqs.get_mut(&nums[lo]).unwrap() -= 1;
                lo += 1;
            }
            max = max.max(hi - lo + 1);
        }

        max as i32
    }
}
